test/phabricator_tests/test_phabricator.py

```diff
@@ -0,0 +1,632 @@
+        response = PhabricatorService().user_query_ids(
+            phids=fake_PHIDS,
+            phab=self.fake_phab
+        )
+
+        # Validate function calls and response
+        self.fake_phab.user.query.assert_called_with(
+            phids=fake_PHIDS
+        )
+        self.assertEqual(expected_response, response)
+
+    def test_user_query_usernames_successful(self):
+        """
+        Tests user_query_ids API Call (successful)
+        """
+        # Set up mock return values and side effects
+        expected_response = "Successful call!"
+        fake_usernames = ["dummy_user_1", "dummy_user_2"]
+
+        # Call function
+        response = PhabricatorService().user_query_usernames(
+            usernames=fake_usernames,
+            phab=self.fake_phab
+        )
+
+        # Validate function calls and response
+        self.fake_phab.user.query.assert_called_with(
+            usernames=fake_usernames
+        )
+        self.assertEqual(expected_response, response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids')
+    def test_author_data_in_rawReponse(self, mock_user_query_ids):
+        """
+        Tests author_data function where author_phid IS IN the raw response
+        """
+        # Set up mock return values and side effects
+        fake_user, fake_raw_response = self.mock.author_data_rawResponse()
+
+        # Call function
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=fake_raw_response,
+            phab=None
+        )
+
+        # Validate function calls and response
+        mock_user_query_ids.assert_not_called()
+        self.assertEqual(fake_user, user)
+        self.assertEqual(fake_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           return_value=["New User!"])
+    def test_author_data_not_in_rawResponse(self, mock_user_query_ids):
+        """
+        Tests author_data function where author_phid IS NOT IN the raw response
+        """
+        # Set up mock return values and side effects
+        expected_raw_response = ["New User!"]
+        expected_user = "New User!"
+        fake_user, _ = self.mock.author_data_rawResponse()
+
+        # Call function
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=[],
+            phab=None
+        )
+
+        # Validate function calls and response
+        mock_user_query_ids.assert_called_with(
+            [fake_user['phid']],
+            None
+        )
+        self.assertEqual(expected_user, user)
+        self.assertEqual(expected_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           side_effect=InternetError)
+    def test_author_data_failure(self, mock_user_query_ids):
+        """
+        Tests author_data function where user_query_ids throws an error (i.e. 404)
```

Comment:
Can you specify when it throws an error?

Reply:
I'm trying to simulate 'APIError' found here: https://github.com/disqus/python-phabricator/blob/ad08e335081531fae053a78a1c708cd11e3e6c49/phabricator/__init__.py